### ISSUE TITLE
Optimizer not defined - missing weight updates for run_eqc.py

### ISSUE THREAD

**Issue by GeorgMiller** [askolik/eqc_for_nco#1]
Hey Andrea, thank you for sharing your code on your very interesting publication! 

I believe you might have a little mistake in the src/model/q_learning.py or in the src/model/tsp_q_learning.py.

When I try executing the run_eqc.py with ```CircuitType.EQC``` (without the use of trainable_scaling and trainable_obs_weight), no optimizer is added to the ```self.optimizer``` list: 

```python
# line 46
        self.optimizer = tf.keras.optimizers.Adam(learning_rate=self.learning_rate, amsgrad=True)
        self.optimizer_output = tf.keras.optimizers.Adam(learning_rate=self.learning_rate_out)
        self.optimizer_input = tf.keras.optimizers.Adam(learning_rate=self.learning_rate_in)

        self.optimizers = []
        self.w_idx = []

        if self.circuit_type == CircuitType.HWE: # in config, circuit_type == CircuitType.EQC
            self.optimizers = [self.optimizer]
            self.w_idx = [0]

        if self.trainable_scaling: # in config, trainable_scaling == False
            self.optimizers.append(self.optimizer_input)
            self.w_idx.append(self.optimizers.index(self.optimizer_input))

        if self.trainable_obs_weight: # in config, trainable_obs_weight== False
            self.optimizers.append(self.optimizer_output)
            self.w_idx.append(self.optimizers.index(self.optimizer_output))

```

However, in  src/model/tsp_q_learning.py, this ```self.optimizers``` list is checked. Since its empty, no gradient updates are performed. 

```python
# line 402
   with tf.GradientTape() as tape:
            tape.watch(self.model.trainable_variables)
            exp_values = self.model([tfq.convert_to_tensor([cirq.Circuit()]*self.batch_size), states])
            exp_val_masks = self.get_masks_for_actions(edge_weights, partial_tours)
            q_values_masked = tf.reduce_sum(tf.multiply(exp_values, exp_val_masks), axis=1)
            loss = self.loss_fun(target_q_values, q_values_masked)

        grads = tape.gradient(loss, self.model.trainable_variables)
       
        if len(self.optimizers) == 1: # len(self.optimizers) == 0
            self.optimizers[0].apply_gradients(zip(grads, self.model.trainable_weights))
        else:
            for optimizer, w in zip(self.optimizers, self.w_idx): # zip is empty
                optimizer.apply_gradients([(grads[w], self.model.trainable_variables[w])])

        return loss.numpy()

```
I modified the code, such that the gradients are applied:

```python
# line 402
    with tf.GradientTape() as tape:
            tape.watch(self.model.trainable_variables)
            exp_values = self.model([tfq.convert_to_tensor([cirq.Circuit()]*self.batch_size), states])
            exp_val_masks = self.get_masks_for_actions(edge_weights, partial_tours)
            q_values_masked = tf.reduce_sum(tf.multiply(exp_values, exp_val_masks), axis=1)
            loss = self.loss_fun(target_q_values, q_values_masked)

        grads = tape.gradient(loss, self.model.trainable_variables)
        self.optimizer.apply_gradients(zip(grads, self.model.trainable_variables))

        return loss.numpy()

```

I tried now various learning rates [0.01, 0.001, 0.0001, 0.00001], but none of them is able to solve the environment of a 10 instance TSP. I was wondering, if maybe some other parameters in the config of the run_eqc.py file need to be changed in order to reproduce the results from you paper? 

Thank you a lot for your time!

**Comment by GeorgMiller** [comment]
Oh and i forgot to mention maybe, that I only changed  'learning_rate_in': 0.00001, to  'learning_rate': 0.00001, as well as the float values of the learning rate in the run_epc.py file.

**Comment by askolik** [comment]
Hi Georg, thanks for reaching out!
Do I understand correctly that you are trying to run the EQC without the "trainable scaling" hyperparameter? In that case, it should indeed not work because the graph weights (your input data) is directly tied to the trainable parameters. If you are not adding that optimizer, there are no trainable parameters in the model, hence the error message you are getting. (This is different for the hardware efficient circuits, which have separate trainable parameters unrelated to the input data.)

**Comment by GeorgMiller** [comment]
Thanks for your quick reply! Indeed, I didnt set the trainable scaling hyperparameter to True (this solves my issue!). 

However I might have a second small question: In your paper you separate between train and validation dataset, but I cant find the validation data in the repository. Is it the same as the ```test``` data? In the ```tsp_10_reduced_train.pickle``` and ```tsp_20_reduced_train.pickle```  there are entries for ```x_test``` and ```y_test``` (additionally to the ```x_train``` and ```y_train```), but in the  ```tsp_5_reduced_train.pickle```  this test data is missing. I also checked the validation_performance files, but couldnt find the validation data itself (but maybe I'm missing it?)

**Comment by askolik** [comment]
Apologies for the confusion, there is an inconsistency in how I stored the data. For all TSP sizes, there are 1000 instances overall in the pickle files. For TSP10 and TSP20, I split those into 100 training, and 900 validation samples (also inconsistently named "test" samples!). I didn't do this for the TSP5, but if you just take the first 100 instances in the pickled list as the training instances you should get the same division of data that I used.

As a bit of additional information you may need: the 100 validation samples that I used are also simply the first ones in the "test" data list. The validation performance files only include the approximation ratios that I computed from the trained models, so that the figures can easily be recreated, but there is otherwise no interesting data there.

**Comment by GeorgMiller** [comment]
Thanks for the clarification!
